File: 프로그래머스/3/60062. 외벽 점검/외벽 점검.cpp
#include <bits/stdc++.h>

using namespace std;

int ret=987654321;
int N;

int fac(int n){
    if(n==1) return 1;
    return n * fac(n-1);
}

vector<int> make(int startIdx, vector<int>& weak){
    vector<int> ret;
    if(startIdx==0) return weak; //0부터시작은 그냥 자기자신임
    
    for(int i=startIdx;i<weak.size();++i){ // [1 5 6 7] 에서 [5 6 7] 넣기
        ret.push_back(weak[i]);
    }
    for(int i=0; i<startIdx;++i){ // 앞에 남은것들 , 1을 1+12해서 넣기
        ret.push_back(weak[i]+N);
    }
    return ret;
}

void testmake(){
    vector<int> vv = {1,5,6,7};
    vector<int> tmp = make(1,vv);
    for(auto i :tmp){
        cout<<i<<" ";
    }
}

int solution(int n, vector<int> weak, vector<int> dist) {

    N=n;
    // cout<<fac(8);
    sort(dist.begin(),dist.end()); // 순열위해 정렬필요
    // int n = weak.size();
    
    do{
        for(int i=0;i<weak.size();++i){ //weak를 순열돌리기 (시작위치 다르게)
            vector<int> weakArr = make(i,weak);
            int idx=0; //dist의 idx == 사람수
            // int i=0; // weak의 idx
            int cur = weakArr[0] + dist[0]; //현재사람의 커버가능한 최대거리
            int flag=1; //가능한지
            for(int j=0;j<weakArr.size();++j){ // j : weak의 idx
                if(weakArr[j] > cur){ //커버할수없는경우, 새사람필요
                    if(idx + 1 == dist.size()){ //새사람이 필요한데, 사람이없으면 실패
                        flag = 0; break;
                    }
                    cur = weakArr[j] + dist[++idx];
                }                
            }
            if(flag) ret = min(ret, idx+1); //인덱스+1 == 사람수
        }
    }while(next_permutation(dist.begin(),dist.end()));
    
    // testmake();
    if(ret==987654321) return -1;
    return ret;
}

File: 프로그래머스/3/60062. 외벽 점검/외벽 점검.java
import java.util.*;

class Solution {
    
    static int N, len, ret = Integer.MAX_VALUE;
    static int[] Weak;
    static Integer[] Dist;
    
    void go(int cnt, int pos, int vis) {
        // 모든 구멍을 메운 경우
        if(vis == (1 << len) - 1) {
            ret = Math.min(ret, cnt);
            return;
        }
        
        // 사람을 다 써버린 경우
        if(cnt >= Dist.length) {
            return;
        }
        
        // 현재 사람이 커버할 수 있는 범위 계산
        int coverage = Dist[cnt];
        int newVis = vis;
        
        // pos부터 시계방향으로 coverage만큼 이동하며 구멍들을 메움
        for(int i = 0; i < len; ++i) {
            int currentPos = (pos + i) % len;
            
            // 거리 계산 (원형이므로 조심)
            int distance;
            if(pos + i < len) {
                distance = Weak[pos + i] - Weak[pos];
            } else {
                distance = Weak[pos + i] - Weak[pos];
            }
            
            if(distance > coverage) {
                break;
            }
            
            newVis |= (1 << currentPos);
        }
        
        // 모든 구멍을 메운 경우
        if(newVis == (1 << len) - 1) {
            ret = Math.min(ret, cnt + 1);
            return;
        }
        
        // 아직 메우지 못한 구멍에서 다음 사람 투입
        for(int i = 0; i < len; ++i) {
            if((newVis & (1 << i)) == 0) { // 아직 메우지 못한 구멍
                go(cnt + 1, i, newVis);
            }
        }
    }
    
    public int solution(int n, int[] weak, int[] dist) {
        N = n;
        len = weak.length;
        ret = Integer.MAX_VALUE;
        
        // 원형을 직선으로 변환 (배열 2배로 확장)
        Weak = new int[len * 2];
        for(int i = 0; i < len; ++i) {
            Weak[i] = weak[i];
            Weak[i + len] = weak[i] + n; // 한 바퀴 돈 것으로 처리
        }
        
        // 이동거리가 큰 순으로 정렬 (그리디)
        Dist = Arrays.stream(dist).boxed().toArray(Integer[]::new);
        Arrays.sort(Dist, (a, b) -> b - a);
        
        // 각 취약점에서 시작해보기
        for(int i = 0; i < len; ++i) {
            go(0, i, 0);
        }
        
        return ret == Integer.MAX_VALUE ? -1 : ret;
    }
}